WalletWasabi/WabiSabi/Models/Decomposition/Decomposition.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Linq;
+using System.Collections.Immutable;
+using NBitcoin;
+using System.Collections.Generic;
+
+namespace WalletWasabi.WabiSabi.Models.Decomposition
+{
+	public sealed record Decomposition : IComparable<Decomposition>, IEquatable<Decomposition>
+	{
+		// Useful for prettifying test assertions, or alternatively as a debug
+		// display attribute:
+		// public override string ToString()
+		// 	=> $"[ {TotalValue} = { string.Join(' ', Outputs) } ]";
+
+		internal Decomposition(params long[] outputValues)
+			: this(outputValues.Select(Money.Satoshis))
+		{
+		}
+
+		internal Decomposition(IEnumerable<Money> outputValues)
+		{
+			Outputs = outputValues.OrderByDescending(x => x).ToImmutableArray();
+			TotalValue = Outputs.Sum();
+		}
+
+		public ImmutableArray<Money> Outputs { get; private init; }
+
+		public Money TotalValue { get; private init; }
+
+		public Decomposition Extend(Money output) =>
+			this with
+			{
+				Outputs = (output <= Outputs[^1])
+					? Outputs.Add(output)
+					: throw new InvalidOperationException("Generated decompositions must be monotonically decreasing"),
+				TotalValue = TotalValue + output
+			};
+
+		public int CompareTo(Decomposition? other)
```

Comment:
This seems to not make sense. Which collection is the bigger one? In terms of items? In terms of total amount? What does this do?

Reply:
I see, so you

1. compare total amount
2. compare length
3. compare lexicographically

I guess it's good if it's useful.